exammode/models.py

```diff
@@ -0,0 +1,107 @@
+from django.db import models
+from django.urls import reverse
+from django.utils import timezone
+
+from course.models import CourseInstance, CourseModule
+from exercise.exercise_models import LearningObject
+from userprofile.models import UserProfile
+
+# Create your models here.
+
+
+class ActiveExamSessionManager(models.Manager):
+
+    def get_queryset(self):
+        initial_queryset = super().get_queryset()
+        queryset = [q for q in initial_queryset if (
+            q.can_start <= timezone.now())]
+
+        return queryset
+
+    def is_active(self):
+        initial_queryset = super().get_queryset()
+        queryset = [q for q in initial_queryset if (q.can_start <= timezone.now() and (
+            timezone.now() <= q.can_start + timezone.timedelta(hours=q.duration)))]
+
+        return queryset
+
+
+class ExamSession(models.Model):
+    """
+    Represents one instance of a course exam
+    """
+
+    course_instance = models.ForeignKey(
+        CourseInstance, on_delete=models.CASCADE)
+    exam_module = models.ForeignKey(
+        CourseModule, on_delete=models.CASCADE, null=True, blank=True)
+    can_start = models.DateTimeField(editable=True, default=timezone.now)
+    duration = models.IntegerField()
+    room = models.CharField(max_length=255, null=True, blank=True)
+
+    objects = models.Manager()
+    active_exams = ActiveExamSessionManager()
+
+    def __str__(self):
+        return " ".join([str(self.course_instance), str(self.can_start)])
+
+    def start_exam(self, user):
+
+        # Checking first if exam content is available. If not, database entries would be pointless
+        learning_objects = LearningObject.objects.filter(
+            course_module__exact=self.exam_module
+        )
+
+        if learning_objects:
+            redirect_url = learning_objects[0].get_display_url()
+
+        else:
+            return reverse("exam_module_not_defined")
+
```

Comment:
The students who currently have an active exam should not be able to start a new exams. Currently the student's exam attempt is removed once the student ends the exam. If this should fail, the attempt stays in the user profile, which will cause problems. This could be fixed by removing exam attemps with some databasequery like:
```
ExamAttempt.objects.filter(exam_taken=self).delete()

Reply:
Removing exam attempts of a closed exam is not implemented.